diff --git a/dist/browser.js b/dist/browser.js
index 6f5e907e9377fc24ed2e01e37e68e1a87b1622c4..64e0e9529119946ead66b65c597d1c36db151be8 100644
--- a/dist/browser.js
+++ b/dist/browser.js
@@ -125,7 +125,8 @@ function RemixBrowser(_props) {
     value: {
       manifest: window.__remixManifest,
       routeModules: window.__remixRouteModules,
-      future: window.__remixContext.future
+      future: window.__remixContext.future,
+      criticalStyles: window.__remixContext.criticalStyles
     }
   }, /*#__PURE__*/React__namespace.createElement(errorBoundaries.RemixErrorBoundary, {
     location: location
diff --git a/dist/components.js b/dist/components.js
index ef21b84685f3d32ea76a35f18da5fdad030b9e20..5024f140fef493e56d0fa1f5dcca9c1a5bd31d2e 100644
--- a/dist/components.js
+++ b/dist/components.js
@@ -193,7 +193,8 @@ function composeEventHandlers(theirHandler, ourHandler) {
 function Links() {
   let {
     manifest,
-    routeModules
+    routeModules,
+    criticalStyles
   } = useRemixContext();
   let {
     errors,
@@ -201,14 +202,16 @@ function Links() {
   } = useDataRouterStateContext();
   let matches = errors ? routerMatches.slice(0, routerMatches.findIndex(m => errors[m.route.id]) + 1) : routerMatches;
   let keyedLinks = React__namespace.useMemo(() => links.getKeyedLinksForMatches(matches, routeModules, manifest), [matches, routeModules, manifest]);
-  return /*#__PURE__*/React__namespace.createElement(React__namespace.Fragment, null, keyedLinks.map(({
+  return /*#__PURE__*/React__namespace.createElement(React__namespace.Fragment, null, [
+    criticalStyles ? /*#__PURE__*/React__namespace.createElement('style', { key: 'style' }, criticalStyles) : null,
+    keyedLinks.map(({
     key,
     link
   }) => links.isPageLinkDescriptor(link) ? /*#__PURE__*/React__namespace.createElement(PrefetchPageLinks, _rollupPluginBabelHelpers["extends"]({
     key: key
   }, link)) : /*#__PURE__*/React__namespace.createElement("link", _rollupPluginBabelHelpers["extends"]({
     key: key
-  }, link))));
+  }, link)))]);
 }
 
 /**
diff --git a/dist/esm/browser.js b/dist/esm/browser.js
index 243b58370be8a61dbc42477e5b7f75e0e695faba..f18abcf2062e43daca29b5f9b21fad91cd690eb9 100644
--- a/dist/esm/browser.js
+++ b/dist/esm/browser.js
@@ -161,7 +161,8 @@ function RemixBrowser(_props) {
     value: {
       manifest: window.__remixManifest,
       routeModules: window.__remixRouteModules,
-      future: window.__remixContext.future
+      future: window.__remixContext.future,
+      criticalStyles: window.__remixContext.criticalStyles
     }
   }, /*#__PURE__*/React.createElement(RemixErrorBoundary, {
     location: location
diff --git a/dist/esm/components.js b/dist/esm/components.js
index 18f36021d36fafc47c85a5803c9e2779e10b22f7..17b055cf7a6d6fd72bad049d3052117add2ff525 100644
--- a/dist/esm/components.js
+++ b/dist/esm/components.js
@@ -169,7 +169,8 @@ function composeEventHandlers(theirHandler, ourHandler) {
 function Links() {
   let {
     manifest,
-    routeModules
+    routeModules,
+    criticalStyles
   } = useRemixContext();
   let {
     errors,
@@ -177,14 +178,16 @@ function Links() {
   } = useDataRouterStateContext();
   let matches = errors ? routerMatches.slice(0, routerMatches.findIndex(m => errors[m.route.id]) + 1) : routerMatches;
   let keyedLinks = React.useMemo(() => getKeyedLinksForMatches(matches, routeModules, manifest), [matches, routeModules, manifest]);
-  return /*#__PURE__*/React.createElement(React.Fragment, null, keyedLinks.map(({
+  return /*#__PURE__*/React.createElement(React.Fragment, null, [
+    criticalStyles ? /*#__PURE__*/React.createElement('style', { key: 'style' }, criticalStyles) : null,
+    keyedLinks.map(({
     key,
     link
   }) => isPageLinkDescriptor(link) ? /*#__PURE__*/React.createElement(PrefetchPageLinks, _extends({
     key: key
   }, link)) : /*#__PURE__*/React.createElement("link", _extends({
     key: key
-  }, link))));
+  }, link)))]);
 }
 
 /**
diff --git a/dist/esm/links.js b/dist/esm/links.js
index 3aa7b5f618a3ef497a2db73d49aeb7492dab41c9..867012f5c705da0fbbade9ea5ba1b6977fd05603 100644
--- a/dist/esm/links.js
+++ b/dist/esm/links.js
@@ -26,15 +26,21 @@ function getKeyedLinksForMatches(matches, routeModules, manifest) {
   let descriptors = matches.map(match => {
     var _module$links;
     let module = routeModules[match.route.id];
-    return ((_module$links = module.links) === null || _module$links === void 0 ? void 0 : _module$links.call(module)) || [];
-  }).flat(1);
+    let route = manifest.routes[match.route.id];
+    return [
+      route.css ? route.css.map(href => ({ rel: "stylesheet", href })) : [],
+      ((_module$links = module.links) === null || _module$links === void 0 ? void 0 : _module$links.call(module)) || [],
+    ];
+  }).flat(2);
   let preloads = getCurrentPageModulePreloadHrefs(matches, manifest);
   return dedupeLinkDescriptors(descriptors, preloads);
 }
-async function prefetchStyleLinks(routeModule) {
-  if (!routeModule.links || !isPreloadSupported()) return;
-  let descriptors = routeModule.links();
-  if (!descriptors) return;
+async function prefetchStyleLinks(route, routeModule) {
+  let descriptors = [
+    route.css ? route.css.map(href => ({ rel: "stylesheet", href })) : [],
+    routeModule.links ? routeModule.links() : []
+  ].flat(1);
+  if (descriptors.length === 0) return;
   let styleLinks = [];
   for (let descriptor of descriptors) {
     if (!isPageLinkDescriptor(descriptor) && descriptor.rel === "stylesheet") {
diff --git a/dist/esm/routes.js b/dist/esm/routes.js
index c33ccae38caa36dc0ca6649592ce8d52652a0677..36d7701ad191812fcd17059ba395022897282fe6 100644
--- a/dist/esm/routes.js
+++ b/dist/esm/routes.js
@@ -121,7 +121,7 @@ function wrapShouldRevalidateForHdr(routeId, routeShouldRevalidate, needsRevalid
 }
 async function loadRouteModuleWithBlockingLinks(route, routeModules) {
   let routeModule = await loadRouteModule(route, routeModules);
-  await prefetchStyleLinks(routeModule);
+  await prefetchStyleLinks(routeModule, routeModule);
   return {
     ...routeModule,
     default: undefined,
diff --git a/dist/links.js b/dist/links.js
index 9565d778da216958e4674ec8258634fcc937a8db..f61d87da47320f19a1164f1ff29f5fb0b929b28b 100644
--- a/dist/links.js
+++ b/dist/links.js
@@ -30,15 +30,21 @@ function getKeyedLinksForMatches(matches, routeModules, manifest) {
   let descriptors = matches.map(match => {
     var _module$links;
     let module = routeModules[match.route.id];
-    return ((_module$links = module.links) === null || _module$links === void 0 ? void 0 : _module$links.call(module)) || [];
-  }).flat(1);
+    let route = manifest.routes[match.route.id];
+    return [
+      route.css ? route.css.map(href => ({ rel: "stylesheet", href })) : [],
+      ((_module$links = module.links) === null || _module$links === void 0 ? void 0 : _module$links.call(module)) || [],
+    ];
+  }).flat(2);
   let preloads = getCurrentPageModulePreloadHrefs(matches, manifest);
   return dedupeLinkDescriptors(descriptors, preloads);
 }
-async function prefetchStyleLinks(routeModule) {
-  if (!routeModule.links || !isPreloadSupported()) return;
-  let descriptors = routeModule.links();
-  if (!descriptors) return;
+async function prefetchStyleLinks(route, routeModule) {
+  let descriptors = [
+    route.css ? route.css.map(href => ({ rel: "stylesheet", href })) : [],
+    routeModule.links ? routeModule.links() : []
+  ].flat(1);
+  if (descriptors.length === 0) return;
   let styleLinks = [];
   for (let descriptor of descriptors) {
     if (!isPageLinkDescriptor(descriptor) && descriptor.rel === "stylesheet") {
diff --git a/dist/routes.js b/dist/routes.js
index 48b804e88d786427b0e1bd98e2cce602d0bdfb85..b2ac00292168775e9d3f5c695544d5a9747bd150 100644
--- a/dist/routes.js
+++ b/dist/routes.js
@@ -142,7 +142,7 @@ function wrapShouldRevalidateForHdr(routeId, routeShouldRevalidate, needsRevalid
 }
 async function loadRouteModuleWithBlockingLinks(route, routeModules$1) {
   let routeModule = await routeModules.loadRouteModule(route, routeModules$1);
-  await links.prefetchStyleLinks(routeModule);
+  await links.prefetchStyleLinks(route, routeModule);
   return {
     ...routeModule,
     default: undefined,
diff --git a/dist/server.js b/dist/server.js
index 57d3da0b0125f7316ea58e0d6a63e3acc134114d..70da457f5de0b8eb0583ac595ff4024ba38c25f9 100644
--- a/dist/server.js
+++ b/dist/server.js
@@ -54,6 +54,7 @@ function RemixServer({
   let {
     manifest,
     routeModules,
+    criticalStyles,
     serverHandoffString
   } = context;
   let routes$1 = routes.createServerRoutes(manifest.routes, routeModules, context.future);
@@ -62,6 +63,7 @@ function RemixServer({
     value: {
       manifest,
       routeModules,
+      criticalStyles,
       serverHandoffString,
       future: context.future,
       serializeError: context.serializeError,