diff --git a/dist/client/client.mjs b/dist/client/client.mjs
index 37c32b4a01d417109380cbdeae902e8c27da78cc..9834577f84c92ac5f0debccd0bd87053e723516c 100644
--- a/dist/client/client.mjs
+++ b/dist/client/client.mjs
@@ -716,8 +716,12 @@ function injectQuery(url, queryToInject) {
     }
     // can't use pathname from URL since it may be relative like ../
     const pathname = url.replace(/#.*$/, '').replace(/\?.*$/, '');
-    const { search, hash } = new URL(url, 'http://vitejs.dev');
-    return `${pathname}?${queryToInject}${search ? `&` + search.slice(1) : ''}${hash || ''}`;
+    const { search, searchParams, hash } = new URL(url, 'http://vitejs.dev');
+    return searchParams.get(queryToInject) === ''
+        ? `${pathname}${search}${hash || ''}`
+        : `${pathname}?${queryToInject}${search ? `&` + search.slice(1) : ''}${
+            hash || ''
+        }`;
 }
 
 export { ErrorOverlay, createHotContext, injectQuery, removeStyle, updateStyle };